### ISSUE TITLE
learnopencv/ALPR/ is not working

### ISSUE THREAD

**Issue by tadam98** [spmallick/learnopencv#658]
[](url
[ALPR_inference_my.ipynb.txt](https://github.com/spmallick/learnopencv/files/8327921/ALPR_inference_my.ipynb.txt)
)

The attached is not working even after corrections. (See my email)

**Comment by sanyam83** [comment]
Hello,
The issue is rectified now and other errors are also rectified, please refer to the notebooks again.
In your case, the folder name "tools" is clashing with another folder with the same name, Hence the error. Try renaming the folder, you will be able to successfully import then.

For example - After renaming the folder,  
`from tools_deepsort import generate_detections as gdet`

**Comment by tadam98** [comment]
Hi, 

You have removed:
```
from google.colab import drive
drive.mount('/content/gdrive')
```
So I guess it is planned to run on my local GPU-Ubuntu-18.04.
I downloaded the complete folder of https://github.com/spmallick/learnopencv/tree/master/ALPR to my local GPU-Ubuntu-18.04.
Started my conda environment that has what is needed (including the requirememts.txt).
With the new ALPR_inference.ipynb the first 7 steps work as documented.
Step [8], first step of the Detector section fails.

%cd ./darknet **fails**.
There is a darknet folder under ./License-plate-detection but I do not think it is the right one.

**Comment by sanyam83** [comment]
If you are downloading the code from here, you need to set paths accordingly like `./License-plate-detection/darknet/` and the darknet folder under this is the right one. Otherwise, if cloning the darknet and other codes like shown in the notebook or the [blog post](https://learnopencv.com/automatic-license-plate-recognition-using-deep-learning), you will not face any errors.

**Comment by tadam98** [comment]
Hi,

OK, I followed [https://learnopencv.com/automatic-license-plate-recognition-using-deep-learning/?ck_subscriber_id=452195442](url)

And decided to try training.
Initial steps are fine.

Then, under Dataset, after "import math" three more imports are needed.
```
import math
import os
import matplotlib.image as image 
import matplotlib.pyplot as plt
```

Make sure the cdw is darknet
`%cd darknet`
Now the images show nice in the plt.

Under "Training" you skipped the location for data.names.
Also the contens are wrong:
# place in darknet/data
classes = 1
```
train = ./darknet/data/obj/train.txt
valid = ./darknet/data/obj/test.txt
names = /content/gdrive/MyDrive/yolov4-darknet/darknet/data/obj.names
backup = ./checkpoint
```
# Change to:
```
classes = 1
train = ./data/obj/train.txt
valid = ./data/obj/test.txt
names = ./data/obj.names
backup = ./checkpoint
```
 You had remains of colab and the path should not have "darknet" in it.

You fogot to mention that "checkdir" should be created under darknet
`!mkdir checkpoint`

You forgot to mention that yolov4.conv.137 should be under ./darknet

Now the traing command works:
`!./darknet detector train data/obj.data cfg/yolov4-obj.cfg yolov4.conv.137 -dont_show -map`

It executed fine and then gave the message below: " Error: cuDNN isn&apos;t found FWD algo for convolution" 

<pre> Tensor Cores are disabled until the first 3000 iterations are reached.
 (next mAP calculation at 1000 iterations) 
 10: -nan, -nan avg loss, 0.000000 rate, 9.552273 seconds, 640 images, 6.814552 hours left
Resizing, random_coef = 1.40 

 512 x 512 
 Error: cuDNN isn&apos;t found FWD algo for convolution.</pre>
 Tensor Cores are disabled until the first 3000 iterations are reached.
 (next mAP calculation at 1000 iterations) 
 10: -nan, -nan avg loss, 0.000000 rate, 9.552273 seconds, 640 images, 6.814552 hours left
Resizing, random_coef = 1.40 

 512 x 512 
 Error: cuDNN isn't found FWD algo for convolution.
[ALPR_inference_my2.ipynb.txt](https://github.com/spmallick/learnopencv/files/8405395/ALPR_inference_my2.ipynb.txt)

I have checked cudnn8 with the nvidia procedure:
```
cd cudnn_samples_v8
cd mnistCUDNN
make clean && make
./mnistCUDNN
```
Executing: mnistCUDNN
cudnnGetVersion() : 8303 , CUDNN_VERSION from cudnn.h : 8303 (8.3.3)
Host compiler version : GCC 9.4.0

There are 1 CUDA capable devices on your machine :
device 0 : sms 68  Capabilities 7.5, SmClock 1650.0 Mhz, MemSize (Mb) 11263, MemClock 7000.0 Mhz, Ecc=0, boardGroupID=0
Using device 0
Resulting weights from Softmax:
0.0000000 0.0000008 0.0000000 0.0000002 0.0000000 1.0000000 0.0000154 0.0000000 0.0000012 0.0000006 
Result of classification: 1 3 5
Test passed!

# summary
I have followed the procedure for training.
Made some corrections to make it start
Something is not working.
will wait a few hours as it appears to be running, but GPU load is at 1-2%.
Sun Apr  3 23:22:57 2022       
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 510.60.02    Driver Version: 512.15       CUDA Version: 11.6     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:0A:00.0  On |                  N/A |
| 25%   33C    P8     7W / 260W |  10141MiB / 11264MiB |      1%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      4258      C   /darknet                        N/A      |
|    0   N/A  N/A     32718      C   /python3.7                      N/A      |
+-----------------------------------------------------------------------------+
```

**Comment by sanyam83** [comment]
For this error :

1. Check if CUDA and cuDNN are of the required versions. According to [darknet](https://github.com/AlexeyAB/darknet) requirements, CUDA >= 10.2 and cuDNN >= 8.0.2 should be installed.
2. Try reducing `subdivisions `in yolov4-obj.cfg file to 32 or 64.

**Comment by tadam98** [comment]
I have cuda 11.0. will check reducing subdivisions as suggested.
________________________________

**Comment by tadam98** [comment]
Increasing the value in ./darknet/cfg/yolov4-obj.cfg file to 32 works ! It is running now.

**Comment by tadam98** [comment]
After 6h it is still at:
```
 Tensor Cores are disabled until the first 3000 iterations are reached.
 (next mAP calculation at 2900 iterations) 
 2820: 0.278743, 0.474088 avg loss, 0.001000 rate, 2.213002 seconds, 180480 images, 2.936276 hours left
Resizing, random_coef = 1.40 

 416 x 416 
 try to allocate additional workspace_size = 82.58 MB 
 CUDA allocate done! 
```
no further progress can be seen.

**Comment by sanyam83** [comment]
These errors are all because memory keeps running out, try increasing the subdivisions more.

**Comment by tadam98** [comment]
Training still "dies" in the middle. I will try it with colab on the larger GPU just to see it training all the way to the end.
I do not have a memory issue when changing to subdivisions=32/64. could be heating, but I saw no indication of this.

**Comment by tadam98** [comment]
Well, it died in the middle on colan with K80 GPU. I changed subdivisions to 32. Running now.
I am also compiling darknet to 2080 TI by uncommenting the correct row in the Makefile to check again on my 2080.

**Comment by tadam98** [comment]
I have successfully downloaded and compiled darknet on **Windows 11, CUDA 11.6 and cuDNN 8.4** using the simple instructions of darknet readme.
```
CUDA-version: 11060 (11060), cuDNN: 8.4.0, GPU count: 1
OpenCV version: 4.5.5
```
Using the settings described hereinabove, based on the guidance in the Notebook, training was successfully completed:
```
Set -points flag:
 `-points 101` for MS COCO
 `-points 11` for PascalVOC 2007 (uncomment `difficult` in voc.data)
 `-points 0` (AUC) for ImageNet, PascalVOC 2010-2012, your custom dataset

 mean_average_precision (mAP@0.50) = 0.897842
Saving weights to ./checkpoint/yolov4-obj_6000.weights
Saving weights to ./checkpoint/yolov4-obj_last.weights
Saving weights to ./checkpoint/yolov4-obj_final.weights
If you want to train from the beginning, then use flag in the end of training command: -clear
```

It could be that my WSL2/Ubuntu 18.04 having CUDA 10.2 has some issues with darkent that used to get stuck. The underlying Windows 11 training is very suitable. (I am keeping CUDA 10.2 chDNN 7.6.5 for Tensorflow 14).

So for now, all is good.
